Update description and error types for `split-by`

drbrain opened the pull request:
# Description

`split-by` only works on a `Record`, the error type was updated to match, and now uses a more-specific type. (Two type fixes for the price of one!)

The `usage` was updated to say "record" as well

# User-Facing Changes

* Providing the wrong type to `split-by` now gives an error messages with the correct required input type

Previously:

```
❯ ls | get name | split-by type
Error:   × unsupported input
   ╭─[entry #267:1:1]
 1 │ ls | get name | split-by type
   ·      ─┬─
   ·       ╰── requires a table with one row for splitting
   ╰────
```

With this PR:

```
❯ ls | get name | split-by type
Error: nu::shell::type_mismatch

  × Type mismatch.
   ╭─[entry #1:1:1]
 1 │ ls | get name | split-by type
   ·      ─┬─
   ·       ╰── requires a record to split
   ╰────
```

# Tests + Formatting

- :green_circle: `toolkit fmt`
- :green_circle: `toolkit clippy`
- :green_circle: `toolkit test`
- :green_circle: `toolkit test stdlib`

# After Submitting

Only generated commands need to be updated

sholderbach commented on crates/nu-command/src/filters/split_by.rs:
```diff
@@ -202,24 +202,18 @@ pub fn data_split(
-                        None,
-                        Vec::new(),
-                    ))
+                    return Err(ShellError::TypeMismatch {
+                        err_message: "requires a record to split".into(),
+                        span,
+                    })
```
Coming from https://github.com/nushell/nushell/issues/10698 maybe `ShellError::OnlySupportsThisInputType` is the better choice here as it also reports the observed wrong type.

drbrain replied:
I’ll update the rust error type

drbrain replied:
Now using `OnlySupportsThisErrorType` and `PipelineEmpty` error types in 695a578a07df382fcbc0c72a89505dd45c4974a4

sholderbach replied:
Thanks! This looks good

sholderbach reviewed (commented):
Thanks for improving the messages here!

fdncred commented:
Wait, split-by doesn't work on tables? I thought all the blah-by commands did?

drbrain commented:
> Wait, split-by doesn't work on tables? I thought all the blah-by commands did?

```
❯ nu --version
0.86.0
❯ ls | split-by type
Error: nu::parser::input_type_mismatch

  × Command does not support table input.
   ╭─[entry #4:1:1]
 1 │ ls | split-by type
   ·      ────┬───
   ·          ╰── command doesn't support table input
   ╰────

❯ ls | describe
table<name: string, type: string, size: filesize, modified: date> (stream)
```

PS: I'm unclear on what the difference between `group-by` and `split-by` in terms of functionality is supposed to be, they seem to be sibling commands that maybe could be merged into one?

fdncred commented:
ya, I've never used `split-by`. By the example it has, it looks like it reads into nested data.
```nushell
> {
        '2019': [
          { name: 'andres', lang: 'rb', year: '2019' },
          { name: 'jt', lang: 'rs', year: '2019' }
        ],
        '2021': [
          { name: 'storm', lang: 'rs', 'year': '2021' }
        ]
    } | split-by lang
  ╭──┬─────────────────╮
  │rb│{record 1 field} │
  │rs│{record 2 fields}│
  ╰──┴─────────────────╯
```
There are other `*-by` commands though. `group-by`, `split-by`, `sort-by`, `uniq-by`. I figured they'd all work on tables, but I guess not. Seems like `split-by` is a bit niche.

amtoine commented:
sounds good :ok_hand: 
if we can get the review threads to close, i'm all for this better error :pray:

fdncred commented:
Looking at the example some more after playing around with input_output_types() in this command, I'm wondering if the only way this command works is if it's a record with tables as values? 

I'd love to see more examples in the help of how this command works.

fdncred reviewed (approved):
Thanks for working on this error.

sholderbach commented on crates/nu-command/src/filters/split_by.rs:
```diff
@@ -202,33 +202,24 @@ pub fn data_split(
-                Vec::new(),
-            ))
-        }
+        _ => return Err(ShellError::PipelineEmpty { dst_span }),
```
If I read the code leading to this point correctly this could still be more than `PipelineData::Empty` e.g. an `ExternalStream`, right?

drbrain replied:
Yes. In 84661a5c20a7fe3c604d80562488513e2e2c2ded I split `PipelineData::Empty` out from the other types which return a `PipelineMismatch`. I looked around and this seems most appropriate as `merge` treats a `ListStream` as a `List` and `ExternalStream` currently can't produce a `Record`.

sholderbach commented on crates/nu-command/tests/commands/split_by.rs:
```diff
@@ -40,7 +49,11 @@ fn errors_if_no_table_given_as_input() {
-        let actual = nu!(
+        let input_mismatch = nu!(cwd: dirs.test(), pipeline("5 | split-by type"));
+
+        assert!(input_mismatch.err.contains("doesn't support int input"));
```
Thanks for covering this as well. This reads like the message you get from the type check in the parse/static check phase in the form of `ParseError::InputMismatch`. But yeah your message improvements should still be relevant as for `get` the type seems unknown at parse time.

fdncred commented:
@drbrain Are you still looking into this?

drbrain commented:
I had a busy week.

It should be ready to go with 84661a5c20a7fe3c604d80562488513e2e2c2ded

sholderbach reviewed (approved):
Awesome thanks, this looks good and well covered
